### DISCUSSION TITLE
Invoking methods with no arguments from open62541 based client

### DISCUSSION THREAD

**Discussion by jknudstr** [FreeOpcUa/opcua-asyncio#1197] [Q&A]
I’m trying to invoke a method in an OPC UA server based on opcua-asyncio.
When trying to invoke a method with no arguments from the open62541 client, defined like this:

   @uamethod
   def RPC_Init(self, parent):
        …

I get the following error:
Traceback (most recent call last):
File …/address_space.py", line 587, in _call
  result = await self._run_method(node.call, method.ObjectId, *method.InputArguments)
TypeError: Value after * must be an iterable, not NoneType
The problem is that when there are no arguments for the method call, open62541 submits “nullptr” for the arguments, which ends up being a “NoneType” in the Python code.
So far, I have found no way (no workaround) to avoid this in the open62541 client.

I should mention that this works with the Softing OPC UA client SDK. I.e., for methods having no arguments, this ends up being an empty list or tuple.
For the purpose of testing, I implemented a small work-around in “asyncua/server/address_space.py” (my modifications marked with “# <<<<<”):

    async def _call(self, method):
        self.logger.info("Calling: %s", method)
        res = ua.CallMethodResult()
        if method.ObjectId not in self._aspace or method.MethodId not in self._aspace:
            res.StatusCode = ua.StatusCode(ua.StatusCodes.BadNodeIdInvalid)
        else:
            node = self._aspace[method.MethodId]
            if node.call is None:
                res.StatusCode = ua.StatusCode(ua.StatusCodes.BadNothingToDo)
            else:
                try:
                    if (method.InputArguments == None): # <<<<<
                        input_args = [] # <<<<<
                    else: # <<<<<
                        input_args = method.InputArguments # <<<<<
                    #result = await self._run_method(node.call, method.ObjectId, *method.InputArguments) # <<<<<
                    result = await self._run_method(node.call, method.ObjectId, *input_args) # <<<<<
        …

whereby it is checked if “method.InputArguments” is “None” and if yes, it is redefined to become an empty list.

Then it works.

My question now is, what is the proper solution to this issue:

1.	Trying to find a work-around in my open62541 based client, or
2.	Trying to have an update introduced in the open62541 SDK to ensure that an ‘array’ is always submitted, or
3.	Implementing the above-mentioned, or similar, fix in “asyncua/server/address_space.py”, or
4.	Or something else…

Thanks in advance for any help to resolve this issue!

**Comment by AndreasHeine** [comment]
the question is how dose the opc ua request look like compared to the other if you investigate with wireshark -> with that we can crosscheck to opc ua spec what it should be!

**Comment by jknudstr** [comment]
OK, thanks for the quick feedback. I will try to produce this information, using Wireshark.

Just for information, I added a print statement in “asyncua/server/address_space.py” in the “_call()” function, to see what is received, in this case from Softing and from my open62541 client:

Softing:
>>>>> _call(): node.call=method.ObjectId=|NodeId(Identifier='MAIN.Actuator1', NamespaceIndex=4, NodeIdType=<NodeIdType.String: 3>)|, method.InputArguments=|[]|

open62541 client:
>>>>> _call(): node.call=method.ObjectId=|NodeId(Identifier='MAIN.Actuator1', NamespaceIndex=4, NodeIdType=<NodeIdType.String: 3>)|, method.InputArguments=|None|

**Reply by AndreasHeine** [reply]
https://reference.opcfoundation.org/v104/Core/docs/Part4/5.11.2/

<html><body>
<a name="Table65">

inputArguments [] | BaseDataType | List of input argument values. An empty list  indicates that there are no input arguments. The size and order of this  list matches the size and order of the input arguments defined by the  input InputArguments Propertyof the Method.                         The name, a description and the data type of each argument are defined by the Argumentstructure in each element of the method’s InputArguments Property.
-- | -- | --


</a>
</body>
</html>inputArguments []
	

BaseDataType
	

List of input argument values. An empty list indicates that there are no input arguments. The size and order of this list matches the size and order of the input arguments defined by the input [InputArguments Property](https://reference.opcfoundation.org/search/17?t=InputArguments%20Property)of the [Method](https://reference.opcfoundation.org/search/17?t=Method).

The name, a description and the data type of each argument are defined by the [Argument](https://reference.opcfoundation.org/search/17?t=Argument)structure in each element of the method’s [InputArguments Property](https://reference.opcfoundation.org/search/17?t=InputArguments%20Property).

**Reply by AndreasHeine** [reply]
as far as spec i would say none is incorrect!

**Reply by AndreasHeine** [reply]
ping @GoetzGoerisch do you know something about it?

**Reply by schroeder-** [reply]
None(length -1) or an empty array(length 0) are both allowed for empty arrays:

> If an [Array](https://reference.opcfoundation.org/search/25?t=Array)is null, then its length is encoded as −1. An [Array](https://reference.opcfoundation.org/search/25?t=Array)of zero length is different from an [Array](https://reference.opcfoundation.org/search/25?t=Array)that is null so encoders and decoders shall preserve this distinction.

**Reply by AndreasHeine** [reply]
@schroeder- this explaines the array topic in general but in the call service its explisitly says an empty list... so thats confusing!
would be interesting what other commercial sdk's do? like prosys / C# / unified automation / ... i mean sofring does it right

**Reply by schroeder-** [reply]
Softing is always sending None for empty arrays. I don't know about other sdks, but from my experience most are sending []. Because I stumbled a cross some issues in some opc ua libraries when using against my OPC UA Server that uses the softing sdk, that always send None for all empty arrays.

**Reply by GoetzGoerisch** [reply]
From above I see that the array shall be empty and as the quote from @schroeder- applies looks like a issue of open62541.
@jknudstr please raise an issue in open62541 on this.

**Reply by jknudstr** [reply]
@GoetzGoerisch: OK, will do.
Thanks all for looking into this!

**Reply by schroeder-** [reply]
I think both is valid. But a lot of sdks have problems with Null arrays. So we should fix our code to accept Null as well.

**Comment by jknudstr** [comment]
Thanks again Andreas.

Would be interesting to know if this is an issue to be fixed in open62541.

In order for us to proceed, we will try to find a work-around in the open62542 client.
If that is not possible, we could maybe install an updated version of opcua-asyncio with the above mentioned work-around.
This is not ideal.
Would you consider to implement this work-around in opcua-asyncio (temporarily)?

Note: We're using the open62541 client with the OPC UA server on Beckhoff Soft-PLCs (TwinCAT).
That works, despite "null" is sent from the client in case the methods have no arguments, so maybe the Beckhoff OPC UA server has implemented the work-around.

**Comment by danmoser** [comment]
Folks, I'm facing exactly the same issue. FYI, [UaExpert](https://www.unified-automation.com/products/development-tools/uaexpert.html) is following the standard (ie., providing an empty list as input argument).

More details at https://github.com/open62541/open62541/issues/5627

**Reply by schroeder-** [reply]
Fixed in #1217

**Reply by jknudstr** [reply]
Sounds good! Many thanks!
